Add a guide to metrics for monitoring Teleport

ptgott opened the pull request:
Closes #40664

This change turns the Metrics guide in `admin-guides` into a conceptual guide to the most important metrics for monitoring a Teleport cluster.

Since Agent metrics have inconsistent comprehensiveness across Teleport services--and to reduce the scope of this change--this guide focuses on self-hosted clusters.

To make this a conceptual guide instead of a reference, this change removes the reference table from the `admin-guides` metrics page. There is already a table in the dedicated metrics reference guide.

Note that, while the new metrics guide is specific to self-hosted clusters, this change does not move the guide to the subsection of Admin Guides related to self-hosting Teleport. Doing this would mean having one subsection of Admin Guides for diagnostics-related guides and one subsection for self-hosted-specific diagnostics, which is potentially confusing. We may also want to add Agent-specific metrics eventually.

Finally, this change does not include alert thresholds for the metrics it describes. We can define these in a subsequent change.

ptgott committed:
Add a guide to metrics for monitoring Teleport

Closes #40664

This change turns the Metrics guide in `admin-guides` into a conceptual
guide to the most important metrics for monitoring a Teleport cluster.

Since Agent metrics have inconsistent comprehensiveness across Teleport
services--and to reduce the scope of this change--this guide focuses on
self-hosted clusters.

To make this a conceptual guide instead of a reference, this change
removes the reference table from the `admin-guides` metrics page. There
is already a table in the dedicated metrics reference guide.

Note that, while the new metrics guide is specific to self-hosted
clusters, this change does not move the guide to the subsection of Admin
Guides related to self-hosting Teleport. Doing this would mean having
one subsection of Admin Guides for diagnostics-related guides and one
subsection for self-hosted-specific diagnostics, which is potentially
confusing. We may also want to add Agent-specific metrics eventually.

Finally, this change does not include alert thresholds for the metrics
it describes. We can define these in a subsequent change.

mmcallister commented on docs/pages/admin-guides/management/diagnostics/metrics.mdx:
```diff
@@ -14,12 +23,168 @@ collect from your Teleport cluster.
+The backend throughput metrics discussed in the previous section map on to
+latency metrics. Whenever the Auth Service increments one of the throughput
+metrics, it reports one of the corresponding latency metrics. See the table
+below for which throughput metrics miap to which latency metrics. Each metric
```
```suggestion
below for which throughput metrics map to which latency metrics. Each metric
```

ptgott replied:
Fixed in b6b8d4f6b3b1e6b6b2c74112181feb84ed8bb5bf

evanfreed commented on docs/pages/admin-guides/management/diagnostics/metrics.mdx:
```diff
@@ -14,12 +23,168 @@ collect from your Teleport cluster.
+If an operation results in an error, the Auth Service adds the `_failed` segment
+to the metric name. For example, successfully creating a record increments the
+`teleport_backend_write_requests_total` metric. If the create operation fails,
+the Auth Service increments `teleport_backend_write_requests_failed_total`
```
As an aside, since I'm often looking for useful metrics as an operator, is there a reason this metric doesn't get surfaced [here](https://goteleport.com/docs/reference/monitoring/metrics/)?  I would expect this to be the full list of metrics surfaced by Teleport.  I've typically had to go spelunking in [metrics.go](https://github.com/gravitational/teleport/blob/e1abc28dcdfabe9010dba4df2ba71d30fbe441a6/metrics.go#L187) to find what I'm looking for since it includes more than that page.

ptgott replied:
We have added to the metrics reference incrementally without any systematic checks for accuracy and completeness. We should auto-generate the reference. I've opened an issue: https://github.com/gravitational/teleport/issues/47000

evanfreed commented on docs/pages/admin-guides/management/diagnostics/metrics.mdx:
```diff
@@ -14,12 +23,168 @@ collect from your Teleport cluster.
+|Delete a range of items|`batch_write_requests`|
+|Update the keepalive status of an item|`write_requests`|
+
+You can use these metrics to define an availability formula, i.e., the
+percentage of reads or writes that succeeded. Take the sum of requests that
+succeeded (including batch requests) over the total sum of requests, multiplied
+by 100. If your backend begins to appear unavailable, you can investigate your
+backend infrastructure.
```
While this does give you calculation of availability, be warned that without calculating for `backend_write_requests_failed_precondition_total` it may be confusing as to why availability is not 100%.  This could be confusing to users.

Sometimes pictures are helpful here's the difference.  I abbreviated the prom query:
`1- (sum(rate(backend_write_requests_failed_total / sum(rate(backend_write_requests_total))`
![Screenshot 2024-09-30 at 3 34 28 PM](https://github.com/user-attachments/assets/86ca56d4-b41f-4856-acca-e5fcd21bff74)
`1- (sum(rate(backend_write_requests_failed_total -sum(rate(teleport_backend_write_requests_failed_precondition_total)) / sum(rate(backend_write_requests_total))`
![Screenshot 2024-09-30 at 3 34 42 PM](https://github.com/user-attachments/assets/8976a2f0-06b6-4063-93eb-e7f687068e1a)

evanfreed commented on docs/pages/admin-guides/management/diagnostics/metrics.mdx:
````diff
@@ -14,12 +23,168 @@ collect from your Teleport cluster.
 
-<Notice scope={["cloud"]} type="tip">
+```text
+sum by (version)(teleport_registered_servers)
````
Just a note, this metric will also count any registered proxies and auths connected.  If someone wanted to count only agents they'd want to subtract their auth/proxies.

ptgott replied:
👍 I've updated this section to focus on versioning, since (a) we already cover Agent resource utilization and (b) there doesn't seem to be a way to group this metric by service

ptgott committed:
Respond to evanfreed feedback

- Describe `backend_write_requests_failed_precondition_total`
- Include the precondition metric in the write availability formula.
- Turn the `registered_servers` discussion into a discussion of Teleport
  instance version, since it's not possible to group this metric by
  service and subtract the count of Auth Service/Proxy Service instances
  from the count of all registered services.

ptgott commented:
@evanfreed I've added new information based on your feedback. Checking to make sure it's accurate. Thanks!

evanfreed commented on docs/pages/admin-guides/management/diagnostics/metrics.mdx:
````diff
@@ -67,10 +67,30 @@ The following backend operation metrics are available:
+```
+1- (sum(rate(backend_write_requests_failed_total -sum(rate(teleport_backend_write_requests_failed_precondition_total)) / sum(rate(backend_write_requests_total))
+```
+
+ef your backend begins to appear unavailable, you can investigate your
````
nit: spelling if instead of ef

evanfreed reviewed (approved):
Looks good, one spelling find
